### PR TITLE
Implement the End Gateway BlockState

### PR DESCRIPTION
**Pull request by Insprill**
# Description
Creates an `EndGatewayMock`.

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Unit tests added.
- [x] Code follows existing code format.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/block/state/EndGatewayMock.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	@Override
+	public @Nullable Location getExitLocation()
+	{
+		return this.exitLocation;
+	}
+
+	@Override
+	public void setExitLocation(@Nullable Location location)
+	{
+		this.exitLocation = location;
```
Comment (syldium) [inline src/main/java/be/seeseemelk/mockbukkit/block/state/EndGatewayMock.java]:
```suggestion
		Preconditions.checkArgument(location == null || Objects.equals(location.getWorld(), isPlaced() ? getWorld() : null), "Cannot set exit location to different world");
		this.exitLocation = location == null ? null : location.toBlockLocation();
```
`Location#toBlockLocation` clones the location, so the state owns this block position.

File: src/main/java/be/seeseemelk/mockbukkit/block/state/EndGatewayMock.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	@Override
+	public @NotNull BlockState getSnapshot()
+	{
+		return new EndGatewayMock(this);
+	}
+
+	@Override
+	public @Nullable Location getExitLocation()
+	{
+		return this.exitLocation;
```
Comment (syldium) [inline src/main/java/be/seeseemelk/mockbukkit/block/state/EndGatewayMock.java]:
```suggestion
		return this.exitLocation == null ? null : this.exitLocation.clone();
```
The location shouldn't be modifiable without using `setExitLocation`.

File: src/test/java/be/seeseemelk/mockbukkit/block/state/EndGatewayMockTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	{
+		assertThrowsExactly(IllegalArgumentException.class, () -> new EndGatewayMock(new BlockMock(Material.BEDROCK)));
+	}
+
+	@Test
+	void constructor_Clone_CopiesValues()
+	{
+		gateway.setExactTeleport(true);
+		gateway.setAge(15L);
+		gateway.setExitLocation(new Location(null, 4, 2, 0));
```
Comment (syldium) [inline src/test/java/be/seeseemelk/mockbukkit/block/state/EndGatewayMockTest.java]:
```suggestion
		gateway.setExitLocation(new Location(this.world, 4, 2, 0));
```
The exit world should be equal to `BlockState#getWorld()`.

**Review by syldium — CHANGES_REQUESTED** [review]
Some nitpicks, since a location is mutable and refers to a world.

**Comment by Insprill** [comment]
> Some nitpicks, since a location is mutable and refers to a world.

Thank you! Looks like I got a bit hasty on this one and didn't do my research.

**Review by thelooter — APPROVED** [review]

**Pull request merged**